[exporter/tanzuobservability] Set a field `source` on the span as per OTLP resources

deosu opened the pull request:
**Description:** Update OTel Collector's Exporter to match WF Proxy Handling of `source`

**Link to tracking Issue:** N/A

**Testing:** Unit Testing & Manual Testing

**Documentation:** N/A

deosu committed:
Update OTel Collector's Exporter to match WF Proxy Handling of `source`

Add Source to span object.

deosu committed:
Update CHANGELOG.md

Update OTel Collector's Exporter to match WF Proxy Handling of source.

deosu committed:
Add default value for source

Update transformer.go to set-up default value for the source.

deosu committed:
Add a README.md and remove duplicate resource attributes from Span.

Update README.md with information about the new attribute `source`.

deosu committed:
Unit tests.

Add unit tests to verify new source tag value.

deosu committed:
Unit tests.

Add unit tests to verify new source tag value.

deosu committed:
Merge branch 'main' into MONIT-26134

deosu committed:
Use semconv constants.

Update transformer.go & an unit test transformer_test.go with semconv constants.

deosu committed:
Merge remote-tracking branch 'origin/MONIT-26134' into MONIT-26134

oppegard commented on exporter/tanzuobservabilityexporter/transformer_test.go:
```diff
@@ -223,6 +223,70 @@ func TestTraceStateTranslatedToTag(t *testing.T) {
+	}
+
+	//TestCase2: source value from resAttrs.host.name when source is not present
+	att.InsertString("host_name", "test_host_name")
```
Do we want to use `hostname` instead of `host_name`, to verify that `host.name` has higher precedence?

deosu replied:
updated

oppegard commented on exporter/tanzuobservabilityexporter/transformer.go:
```diff
@@ -86,12 +88,30 @@ func (t *traceTransformer) Span(orig pdata.Span) (span, error) {
+	for _, key := range candidateKeys {
+		if value, isFound := attributes.Get(key); isFound {
+			source = value.StringVal()
+			attributes.Delete(key)
```
Looking at the [Data Ownership](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter#data-ownership) section from the Collector's Exporter docs, we shouldn't be modifying any of the data in `pdata.Traces`. So, we should not call `attributes.Delete()`. Let's find another way to do this, such as by returning the source and resources attributes as tags. The signature of `getSource` might change to:
```go
func getSourceAndResourceTags(attributes pdata.AttributeMap) (string, map[string]string)
```

Then we combine the resource tags with the span tags.

deosu replied:
Creating a copy before we remove.

oppegard commented on exporter/tanzuobservabilityexporter/transformer.go:
```diff
@@ -164,6 +184,11 @@ func attributesToTags(attributes ...pdata.AttributeMap) map[string]string {
 		att.Range(extractTag)
 	}
 
+	if value, isFound := tags["source"]; isFound {
```
Feels like we should pull out a constant of `labelSource = "source"` to `exporter.go`, and use that instead of `"source"` everywhere.

deosu replied:
Done

deosu committed:
Do not update pdata.Traces.

Create a copy of pdata.Traces before removing duplicate attributes.

oppegard commented on exporter/tanzuobservabilityexporter/transformer.go:
```diff
@@ -95,21 +95,24 @@ func (t *traceTransformer) Span(orig pdata.Span) (span, error) {
+func getSourceAndResourceTags(attributes pdata.AttributeMap) (string, pdata.AttributeMap) {
 	candidateKeys := []string{"source", conventions.AttributeHostName, "hostname", conventions.AttributeHostID}
 
+	//attributesWithoutSource := attributes
```
Delete comment 🙂

deosu replied:
My bad :(

deosu committed:
Introduce a constant for source.

Introduce a constant of labelSource = "source" to exporter.go, and use that instead of "source" everywhere.

oppegard commented on exporter/tanzuobservabilityexporter/transformer_test.go:
```diff
@@ -223,6 +224,73 @@ func TestTraceStateTranslatedToTag(t *testing.T) {
+	inNanos := int64(50000000)
+
+	//TestCase1: default value for source
+	att := pdata.NewAttributeMap()
```
Can we give `att` a more descriptive name throughout this test? Like `resAttrs`?

deosu replied:
done

oppegard commented on exporter/tanzuobservabilityexporter/transformer_test.go:
```diff
@@ -223,6 +224,73 @@ func TestTraceStateTranslatedToTag(t *testing.T) {
+	assert.Equal(t, "test_source", actual.Source)
+	assert.Equal(t, "test_host.name", actual.Tags[conventions.AttributeHostName])
+	if value, isFound := actual.Tags[labelSource]; isFound {
+		t.Logf("Tag Source with value " + value + " not expected.")
```
Does `t.logF()` cause a test failure? If so, 👍 . If not, I feel the test should fail if the tag hasn't actually been deleted.

deosu replied:
Added t.Fail(), thank you for pointing this out.

oppegard reviewed (changes requested):
A couple small nits, and then I think we're good!

deosu committed:
Fix unit test.

Add t.Fail() to make sure test fails in case of an unexpected behaviour.

deosu commented on exporter/tanzuobservabilityexporter/transformer.go:
```diff
@@ -86,12 +88,32 @@ func (t *traceTransformer) Span(orig pdata.Span) (span, error) {
 	}, nil
 }
 
+func getSourceAndResourceTags(attributes pdata.AttributeMap) (string, pdata.AttributeMap) {
+	candidateKeys := []string{labelSource, conventions.AttributeHostName, "hostname", conventions.AttributeHostID}
+
+	attributesWithoutSource := pdata.NewAttributeMap()
+	attributes.CopyTo(attributesWithoutSource)
+	var source string
+
+	for _, key := range candidateKeys {
+		if value, isFound := attributesWithoutSource.Get(key); isFound {
+			source = value.StringVal()
+			attributesWithoutSource.Delete(key)
+			break
+		}
+	}
+
+	//returning an empty source is fine as wavefront.go.sdk will set it up to a default value(os.hostname())
+	return source, attributesWithoutSource
+}
```
Looking at the [Data Ownership](https://github.com/open-telemetry/opentelemetry-collector/tree/main/exporter#data-ownership) section from the Collector's Exporter docs, we shouldn't be modifying any of the data in `pdata.Traces`. That is why we are cloning the attributes and then removing `source` from it as we have already added that information in the `source` field. Just to reduce duplicate data.

deosu replied:
Does this sound good?

deosu replied:
@bogdandrutu can you please take a look?

deosu replied:
I have made a few changes to avoid copying AttributeMap. Please review.

deosu commented:
Hello Team, we do not think that the failing test is related to our changes. Please let us know otherwise. 
https://github.com/open-telemetry/opentelemetry-collector-contrib/runs/5223042919?check_suite_focus=true

deosu committed:
Avoid copying AttributeMap.

Avoid copying AttributeMap to reduce memory usage.

deosu committed:
Avoid copying AttributeMap.

Avoid copying AttributeMap to reduce memory usage.

deosu committed:
Fix Lint failure

Remove unnecessary use of if.

deosu committed:
Fix Lint failure

Fix import issues.
